jimp/index.d.ts

```diff
@@ -0,0 +1,207 @@
+// Type definitions for jimp 0.2
+// Project: https://github.com/oliver-moran/jimp#readme
+// Definitions by: Jack Works <https://github.com/Jack-Works>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+/// <reference types="node" />
+
+declare namespace jimp {
+	type Callback<T> = (this: JimpImage, err: Error, data: T) => void;
+	type ResizeMode = string;
+	type PresetFont = string;
+	type FileMINE = string;
+	type AlignMode = number;
+	type FilterType = number;
+
+	// Used in font info provided by parse-bmfont-ascii
+	type FontBoolean = 0 | 1;
+	interface FontChar {
+		id: number;
+		x: number;
+		y: number;
+		weight: number;
+		height: number;
+		xoffset: number;
+		yoffset: number;
+		xadvance: number;
+		page: number;
+		chnl: number;
+	}
+	type Font = {
+		common: {lineHeight: number, base: number, scaleW: number, scaleH: number, pages: number, packed: number},
+		info: {
+			face: string,
+			size: number,
+			bold: FontBoolean,
+			italic: FontBoolean,
+			charset: string,
+			unicode: FontBoolean,
+			stretchH: number,
+			smooth: FontBoolean,
+			aa: number,
+			padding: [number, number, number, number],
+			spacing: [number, number]
+		},
+		kernings: any,
+		pages: JIMP[],
+		chars: {
+			[char: number]: FontChar,
+			[char: string]: FontChar
+		}
+	} | PresetFont;
+
+
+	class JimpImage {
+		constructor(width: number, height: number, callback?: Callback<JimpImage>);
+		constructor(width: number, height: number, initialColor: number, callback?: Callback<JimpImage>);
+
+		/* Get/set meta data */
+		hash(n?: number): string;
+
+		getExtension(): string;
+		getMIME(): FileMINE;
+
+		getPixelColor(x: number, y: number): number;
+		setPixelColor(hex: number, x: number, y: number): void;
+
+		/* Save */
+		write(path: string, callback?: Callback<undefined>): void;
+		getBase64(mine: FileMINE, callback?: Callback<string>): void;
+		getBuffer(mine: FileMINE, callback?: Callback<Buffer>): void;
+
+		/* Parameter */
+		quality(n: number): JimpImage;
+		rgba(is: boolean): JimpImage;
+		filterType(type: FilterType): JimpImage;
+		deflateLevel(level: number): JimpImage;
+
+		/* Operation */
+		clone(): JimpImage;
+		print(font: Font, x: number, y: number, text: string, width?: number): JimpImage;
+
+		/* Adv */
+		color(
+			params: Array<{
+				apply: 'lighten' | 'brighten' | 'darken' | 'desaturate' | 'saturate' | 'greyscale' | 'spin' | 'hue' | 'tint' | 'shade' | 'red' | 'green' | 'blue',
+				params: [number]
+			} | {
+				apply: 'mix',
+				params: [string, number]
+			} | {
+				apply: 'xor',
+				params: [string]
+			}>): JimpImage;
+
+		convolution(matrix: number[][]): JimpImage;
+		scan(x: number, y: number, width: number, height: number, callback?: (this: JimpImage, x?: number, y?: number, idx?: number) => void): JimpImage;
```

Comment:
Optional callback parameters [are senseless](https://github.com/Microsoft/TypeScript-Handbook/blob/master/pages/declaration%20files/Do's%20and%20Don'ts.md#optional-parameters-in-callbacks).